src/huggingface_hub/hf_api.py

```diff
@@ -9182,6 +9158,84 @@ def _prepare_folder_deletions(
             if relpath_to_abspath[relpath] != ".gitattributes"
         ]
 
+    def _prepare_upload_folder_additions(
+        self,
+        folder_path: Union[str, Path],
+        path_in_repo: str,
+        allow_patterns: Optional[Union[List[str], str]] = None,
+        ignore_patterns: Optional[Union[List[str], str]] = None,
+        repo_type: Optional[str] = None,
+        token: Union[bool, str, None] = None,
+    ) -> List[CommitOperationAdd]:
+        """Generate the list of Add operations for a commit to upload a folder.
+
+        Files not matching the `allow_patterns` (allowlist) and `ignore_patterns` (denylist)
+        constraints are discarded.
+        """
+        headers = self._build_hf_headers(token=token)
+
+        folder_path = Path(folder_path).expanduser().resolve()
+        if not folder_path.is_dir():
+            raise ValueError(f"Provided path: '{folder_path}' is not a directory")
+
+        # List files from folder
+        relpath_to_abspath = {
+            path.relative_to(folder_path).as_posix(): path
+            for path in sorted(folder_path.glob("**/*"))  # sorted to be deterministic
+            if path.is_file()
+        }
+
+        # Filter files
+        # Patterns are applied on the path relative to `folder_path`. `path_in_repo` is prefixed after the filtering.
+        filtered_repo_objects = list(
+            filter_repo_objects(
+                relpath_to_abspath.keys(), allow_patterns=allow_patterns, ignore_patterns=ignore_patterns
+            )
+        )
+
+        prefix = f"{path_in_repo.strip('/')}/" if path_in_repo else ""
+
+        # If updating a README.md file, make sure the metadata format is valid
+        # It's better to fail early than to fail after all the files have been uploaded or hashed.
+        for relpath in filtered_repo_objects:
+            if relpath == "README.md":
```

Comment:
And then, instead of the `for ... in ...: if ... == "README.md": .... ... ... break` logic, could you use `relpath_to_abspath` instead like this:

```py
if "README.md" in filtered_repo_objects:
    self._validate_yaml(
        content=relpath_to_abspath["README.md"].read_text(),
        repo_type=repo_type,
        token=token,
    )
```

This way, it's not even needed to build the `CommitOperationAdd` just to read the file.

Reply:
Thanks, much better to use `read_text` from pathlib and avoid the for loop.